include/deal.II/base/mutable_bind.h

```diff
@@ -0,0 +1,320 @@
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_base_mutable_bind_h
+#define dealii_base_mutable_bind_h
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/base/patterns.h>
+#include <deal.II/base/std_cxx14/utility.h>
+#include <deal.II/base/std_cxx17/tuple.h>
+
+#include <tuple>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace Utilities
+{
+  /**
+   * A mutable version of std::bind, that binds all arguments of a function
+   * pointer to a stored tuple, and allows you to update the tuple between
+   * calls.
+   *
+   * An example usage of this class is through the helper function
+   * mutable_bind() that creates a MutableBind object on the fly, based on its
+   * arguments:
+   *
+   * @code
+   * void my_function(const int &a, const double &b);
+   *
+   * auto bound = mutable_bind(my_function, 1, 2.0);
+   *
+   * bound(); // will execute my_function(1, 2.0);
+   *
+   * bound.set_parameters(2, 3.0);
+   * bound(); // will execute my_function(2, 3.0);
+   *
+   * bound.parse_parameters("3: 4.0");
+   * bound(); // will execute my_function(3, 4.0);
+   * @endcode
+   *
+   * The arguments are copied to the tuple, with their reference and const
+   * attributes removed. Only copy constructible objects are allowed as
+   * function arguments. If you need to keep some references around, you may
+   * wrap your function into a std::function, and capture your reference:
+   *
+   * @code
+   * Triangulation<dim> &tria;
+   * std::function<void(double, double)> cube =
+   *    [&tria] (double left, double right) {
+   *    GridGenerator::hyper_cube(tria, left, right);
+   * }
+   *
+   * auto bound_cube = mutable_bind(cube);
+   *
+   * bound_cube.parse_parameters("0: 1");
+   * bound_cube(); // calls GridGenerator::hyper_cube(tria, 0, 1);
+   * @endcode
+   *
+   * @authors Luca Heltai, Matthias Maier, 2019.
+   */
+  template <typename ReturnType, class... FunctionArgs>
+  struct MutableBind
+  {
+    /**
+     * A typedef to the stored std::tuple type.
+     */
+    using TupleType = std::tuple<typename std::remove_cv<
+      typename std::remove_reference<FunctionArgs>::type>::type...>;
+
+    /**
+     * An std::function that stores the the original function.
+     */
+    const std::function<ReturnType(FunctionArgs...)> function;
+
+    /**
+     * Currently stored arguments. These are forwarded to the function object
+     * above, when calling operator()().
+     */
+    TupleType arguments;
+
+    /**
+     * Call the original function, passing as arguments the elements of the
+     * stored TupleType.
+     */
+    ReturnType
+    operator()() const;
+
+    /**
+     * Set the arguments to use in @p function, for next time
+     * operator()() is called.
+     */
+    void
+    set_arguments(const TupleType &value);
```

Comment:
You could probably optimize a few things if you also added the following function:
```
    void
    set_arguments(TupleType &&value) { arguments = std::move(value); }
```

Reply:
Also, I would call the function argument `arguments` to make their intent clear.